osu.Framework/Input/StateChanges/IMouseInput.cs

```diff
@@ -0,0 +1,13 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Framework.Input.StateChanges
+{
+    public interface IMouseInput : IInput
+    {
+        /// <summary>
+        /// Whether this input is performed from a primary touch input source.
+        /// </summary>
+        public bool FromTouchSource { get; set; }
```

Comment:
I catch myself doing this often since C# 8 landed too, but the `public` modifier on the interface member is implied:

```suggestion
        bool FromTouchSource { get; set; }
```

Reply:
Wait..., when did I put this o_O

Reply:
This should not be in the mouse state, as discussed elsewhere.

Reply:
It is required to be in the mouse state, such that `IRequireHighFrequencyMousePosition` would work correctly and also useful for consumers to know whether the current mouse state is as-is because of touch input.

Reply:
Sounds like we should be fixing the clone logic rather than hacking around it. this would imply that *any* deriving classes would be lost in the flow, so by not fixing but working around the issue, it is being left open for future failure (the classes should be sealed if you go down the path you are, but I don't believe this is what we want).

Reply:
I'm not sure what you mean by "the clone logic", and "this would imply that any deriving classes would be lost in the flow", can you elaborate further?

Reply:
`MouseMoveEvent` is not sealed. It could be derived and cause the same loss-of-information you are implying above.

The change in this PR is hacking around an issue without resolving the underlying issue.

Fixing the clone logic implies fixing the case where this loss-of-information occurs, which you suggest is in the IRequireHighFrequency code path. Construction of the base class should be replaced with clone logic that respects the original type and content.

Reply:
I'd also like to see this implemented as a property rather than an interface. It may be slightly harder to implement but I think it's worthwhile... If we want to add additional properties (such as perhaps the touch source, if we maybe allow multiple touches to do this in the future, etc), I wouldn't want to pollute MouseEvent with that sort of information.

If the only issue is the high-frequency path, I think may be fine to store the property in the state (considering syncing...), and then switch on that state for the event to propagate in the hf path.

Reply:
Did you mean a property in an interface?

Reply:
No. I meant keep the property inside `MouseState` as it currently is, and have the interface for UI events.

Reply:
To clarify, you're saying you believe it *should* be in `MouseState` as per this PR?

Reply:
I think it makes sense from a syncing perspective, yes.

Reply:
Are you okay with it being `private` or `internal`, in that case?

Reply:
Also can you explain why you're against keeping it isolated to an interface? This would still allow further properties to be added in the future for the use cases you mention.

To build on my last comment, the user should not be aware of this bool unless they are very specifically wanting to use it (in an already-touch-supporting application). By exposing it in a mouse event, it's not only polluting cross-polluting separate input systems, but I think it's quite misleading – and badly guiding – consumers.

Reply:
Oh so this would mean even if it's gonna pollute the framework implementation internally, it's fine as long as it's not polluted to what the user sees, didn't really think you would be fine but talking about it from this perspective, that makes absolute sense, will change it.

Reply:
> No. I meant keep the property inside MouseState as it currently is, and have the interface for UI events.

I already mentioned having an interface for UI events.

As for `MouseState`, another alternative may be to not store the bool inside `MouseState` but inside `InputState` itself, as a "LastMovementSource" enum value (`Touch`/`Mouse`).

Reply:
Please wait until the discussion has concluded, as a direction forward has not been attained yet.

Reply:
@smoogipoo my criteria for implementation:

- should not be visible to the user via a `Mouse(Anything)` class to consumers
- should be accessible by the user via an interface (for special cases where the user is already supporting touch)

can you agree with that direction?

Reply:
I agree with that in the context of UI events (`Mouse(Anything)`). Not sure about `InputState`, I think it'd be best to expose it in some way - there are still cases where we handle input manually via `InputManager.CurrentState`, and some of those cases _may_ want to know which device the current position comes from.

E.g. if a user wants to handle _mouse_ input only, but accesses `CurrentState.Mouse.Position`, how can they be sure that that position comes from a mouse and not a touch?

Reply:
"handle mouse only" should only be a scenario if the user is also handling touch, at which point they should be using the interface (it would be documented under the touch wiki page or otherwise).

there should be no scenario a consumer supports mouse and only mouse, disabling touch, unless they are aware of touch already.

Reply:
And what if you can't use the interface? For whatever reason - maybe you haven't even received a movement event because your drawable's too new to the hierarchy, or hasn't been moused over.

Example: A cursor that doesn't reside in a 100% screen area, which is shown only during mouse input, that wants to show the current mouse position when it's out of bounds, or an accurate position when no movement event has otherwise been received.

Reply:
Ah, maybe I misunderstood – the goal here was not just to interface the events, but also the state, and have a `TouchBasedMouseState : MouseState`, along with updating the clone/transfer logic to allow for this.

Therefore the following would be possible:

`if (GetContainingInputManager().CurrentState.MouseState is ITouchSourcedMouseState touchSpecificData) { ... }`

Reply:
I see. That sounds difficult to handle, since a lot of that part of code is exposed to the user (see `InputManager.CreateInitialState()`). Beyond the difficulty of transferring states, just deciding when it's actually safe to switch the states around sounds non-trivial. If the user couldn't specify their own mouse state at construction time, then it'd be pretty simple.

I don't know how I feel about the entire `MouseState` member changing, conceptually, and potentially flip-flopping between two types for what is essentially almost a singleton. For what it's worth, I'm also not a fan of my idea of `InputState` having a mutable property either...

Reply:
What do you think about sealing all state classes (see original comment https://github.com/ppy/osu-framework/pull/3675#discussion_r448670134) then working out a solution with that restriction in mind? As I mentioned, it's open to future error if we don't fix this here.

Reply:
I agree with sealing.

Am also thinking if there's any chance that maybe we _don't_ mutate the `MouseState` on a touch-to-mouse event? It would give the user the guarantee that `MouseState` always corresponds to a mouse device, and `TouchState` always corresponds to a touch device, but there'd need to be more extra considerations for `MouseEvent`, like perhaps instead of `IsPressed()` checking `CurrentState.Mouse` it'd check the union of Mouse+Touch (or even another state like an `internal` `InputState.TouchSourcedMouseState`, `internal` is fine in this case because a user can always reference `Mouse` and `Touch` independently to retrieve true device states).

Reply:
Sounds like something that could work, but would require a good amount of time and effort. going the interface way sounds easier for now to fix the issues this PR is for.

Reply:
@peppy About adding interfaces for mouse events to indicate they are from a touch source, there does not seem to be a way to properly organize the logic of whether to construct `TouchBasedMouseXXXEvent` or the regular `MouseXXXEvent`. 

The best I can do is adding a `static MouseXXXEvent Create(InputState, ...)` method to each constructible `MouseEvent`, does that seem fine or do you have a better plan?